appinventor/blocklyeditor/src/blocklyeditor.js

```diff
@@ -514,79 +1085,128 @@ Blockly.ai_inject = function(container, workspace) {
-   * @returns {Function}
-   */
-  function commentRenderer(comment) {
-    return function() {
-      comment.setVisible(comment.visible);
-    }
-  }
+  workspace.injecting = false;
+  workspace.injected = true;
+  workspace.notYetRendered = true;
+  return workspace;
+};
 
-  for (var i = blocks.length - 1; i >= 0; i--) {
-    var block = blocks[i];
-    block.initSvg();
-    block.rendered = true;
-    if (block.disabled && block.updateDisabled) {
-      block.updateDisabled();
-    }
-    if (!isNaN(block.x) && !isNaN(block.y)) {
-      var xy = block.getRelativeToSurfaceXY();
-      block.getSvgRoot().setAttribute('transform',
-        'translate(' + block.x + ',' + block.y + ')');
-      block.moveConnections_(block.x - xy.x, block.y - xy.y);
-    }
-    if (block.comment && block.comment.visible && block.comment.setVisible) {
-      setTimeout(commentRenderer(block.comment), 1);
-    }
+/**
+ * Inject a previously constructed workspace into the designated
+ * container. This implementation is adapted from Blockly's
+ * implementation and is required due to the fact that browsers such as
+ * Firefox do not initialize SVG elements unless they are visible.
+ *
+ * @param {!Element|string} container
+ * @param {!Blockly.WorkspaceSvg} workspace
+ */
+Blockly.ai_inject = function(container, workspace) {
+  Blockly.common.setMainWorkspace(workspace);  // make workspace the 'active' workspace
+  workspace.fireChangeListener(new AI.Events.ScreenSwitch(workspace.projectId, workspace.formName));
+  var gridEnabled = top.BlocklyPanel_getGridEnabled && top.BlocklyPanel_getGridEnabled();
+  var gridSnap = top.BlocklyPanel_getSnapEnabled && top.BlocklyPanel_getSnapEnabled();
+  // Note that in current code, workspace.injected is always true.  Before the 2023 Blockly update
+  // it was possible for workspace.injected to be false, but that is no longer the case.
+  if (workspace.injected) {
+    workspace.setGridSettings(gridEnabled, gridSnap);
+    // Update the workspace size in case the window was resized while we were hidden
+    setTimeout(function() {
+      goog.array.forEach(workspace.blocksNeedingRendering, function(block) {
+        workspace.getWarningHandler().checkErrors(block);
+        block.render();
+      });
+      workspace.blocksNeedingRendering.splice(0);  // clear the array of pending blocks
+      workspace.resizeContents();
+      Blockly.svgResize(workspace);
+      if (workspace.notYetRendered) {
+        workspace.notYetRendered = false;
+        workspace.scrollCenter();
+      }
+      //AI.Blockly.navigationController.enable(workspace);
+    });
+    return;
   }
-  workspace.render();
-  // blocks = workspace.getTopBlocks();
+  // The following code used to be reachable, but currently is not, because workspace.injected is
+  // now (since the 2023 Blockly update) always true.  We are leaving it here, but commented, in case
+  // it turns out that we have to revert to the old behavior.
+  //
```

Comment:
If it's always true that workspace.injected is true, then we should also consider inlining the code in the `if` statement above.

Reply:
If by inlining you mean just getting rid of the if, then sure.  I was just being a bit paranoid and thinking that we might later discover that `workspace.injected` was not always true.  If we inline the code, then we might as well also delete the commented out code.